File: 2021/3/partTwo.ts
import { readFileSync } from "fs";

async function main() {
  let result = 0;
  const file = readFileSync(process.argv[2], "utf-8");
  const rows = file.split("\n").map((r) => {
    const row = r;
    return r;
  });

  let oxygenGenRating = [...rows];
  let waterRating = [...rows];

  const l = rows[0].length;

  for (let i = 0; i < l; i++) {
    let oones = 0;
    let ozeros = 0;
    let wones = 0;
    let wzeros = 0;
    oxygenGenRating.forEach((r, index) => {
      if (parseInt(r.charAt(i)) === 1) {
        oones++;
      } else {
        ozeros++;
      }
    });
    waterRating.forEach((r, index) => {
      if (parseInt(r.charAt(i)) === 1) {
        wones++;
      } else {
        wzeros++;
      }
    });
    if (oxygenGenRating.length > 1) {
      if (oones >= ozeros) {
        oxygenGenRating = oxygenGenRating.filter(
          (r) => parseInt(r.charAt(i)) === 1
        );
      } else if (oones < ozeros) {
        oxygenGenRating = oxygenGenRating.filter(
          (r) => parseInt(r.charAt(i)) === 0
        );
      }
    }
    if (waterRating.length > 1) {
      if (wones >= wzeros) {
        waterRating = waterRating.filter((r) => parseInt(r.charAt(i)) === 0);
      } else if (wones < wzeros) {
        waterRating = waterRating.filter((r) => parseInt(r.charAt(i)) === 1);
      }
    }
  }

  result = parseInt(oxygenGenRating[0], 2) * parseInt(waterRating[0], 2);

  console.log(result);
}

main();

File: 2021/3/part-two.py
import sys

if __name__ == "__main__":
    result = 0
    with open(sys.argv[1], encoding="utf-8") as f:
        rows = [x.rstrip() for x in f]

    oxygen, water = rows, rows

    l = len(rows[0])
    for i in range(0, l):
        if len(oxygen) == 1:
            break
        values = [0, 0]
        for o in oxygen:
            values[int(o[i])] += 1
        if values[0] <= values[1]:
            oxygen = list(filter(lambda r: r[i] == "1", oxygen))
        else:
            oxygen = list(filter(lambda r: r[i] == "0", oxygen))

    for i in range(0, l):
        if len(water) == 1:
            break
        values = [0, 0]
        for w in water:
            values[int(w[i])] += 1
        if values[0] > values[1]:
            water = list(filter(lambda r: r[i] == "1", water))
        else:
            water = list(filter(lambda r: r[i] == "0", water))

    result = int(oxygen[0], 2) * int(water[0], 2)

    print(result)
